src/main/java/me/jellysquid/mods/lithium/common/world/noise/SimplexNoiseCache.java

```diff
@@ -0,0 +1,130 @@
+        this.values = new float[CACHE_SIZE];
+    }
+
+    /**
+     * Attempt to get the cached distance factor, saving computation time.
+     */
+    private float getDistanceFactor(int x, int z) {
+        // Hash key and get index
+        long key = ChunkPos.toLong(x, z);
+        int idx = (int) HashCommon.mix(key) & this.mask;
+
+        if (this.keys[idx] == key) {
+            // Cache hit, return cached value
+            return this.values[idx];
+        }
+
+        // Cache miss, compute and store value
+
+        // A marker for no value.
+        float value = -1.0F;
+
+        long lx = x;
+        long lz = z;
+        long distanceFromOriginSq = lx * lx + lz * lz;
+
+        // Ensure we are 64 grid cells away from the origin.
+        if (distanceFromOriginSq > 64 * 64) {
+            // Reduce the number of island-forming grid cells by sampling noise with a threshold
+            if (this.sampler.sample(x, z) < -0.9) {
+                // Generate a pseudo-random value from 9 to 21
+                value = (MathHelper.abs(x) * 3439.0F + MathHelper.abs(z) * 147.0F) % 13.0F + 9.0F;
+            }
+        }
+
+        // Store values in cache
+        this.keys[idx] = key;
+        this.values[idx] = value;
+
+        return value;
+    }
+
+    /**
+     * Mapped and cleaned up implementation of the End biome source's sampler. Tries to use cached values wherever possible.
+     */
+    public float getNoiseAt(int x, int z) {
+        // [VanillaCopy] TheEndBiomeSource#getNoiseAt
+
+        int gridX = x / GRID_SIZE;
+        int gridZ = z / GRID_SIZE;
+
+        // This is the "center point", offset to center around the current grid cell
+        int gridOriginX = x % GRID_SIZE;
+        int gridOriginZ = z % GRID_SIZE;
+
+        // Initialize density for the central island
+        float density = ISLAND_RADIUS - MathHelper.sqrt(x * x + z * z) * 8.0F;
+
+        // Iterate through 25x25 grid cells
+        for (int offsetX = -12; offsetX <= 12; ++offsetX) {
+            for (int offsetZ = -12; offsetZ <= 12; ++offsetZ) {
+                int globalGridX = gridX + offsetX;
+                int globalGridZ = gridZ + offsetZ;
+
+                // Try to retrieve values from cache
+                float distanceFactor = getDistanceFactor(globalGridX, globalGridZ);
+                if (distanceFactor != -1.0F) {
+                    // Compute the distance to the origin
+                    float deltaX = gridOriginX - offsetX * GRID_SIZE;
+                    float deltaZ = gridOriginZ - offsetZ * GRID_SIZE;
+
+                    // Calculate the density at this grid cell
+                    float scaledDistance = MathHelper.sqrt(deltaX * deltaX + deltaZ * deltaZ) * distanceFactor;
+                    float densityHere = ISLAND_RADIUS - scaledDistance;
+
+                    // Try to return early if we're over the max
+                    if (densityHere > density) {
+                        if (density >= MAX) {
```

Comment:
Shouldn't this be `densityHere >= MAX`

Reply:
Yeah, good catch! It shouldn't have caused any problems, but that shaves off another grid cycle 😅